File: TripKitUI/helper/Categories/UIColor+Variations.m
//
//  UIColor+Variations.m
//  TripPlanner
//
//  Created by Adrian Schoenig on 21/02/13.
//
//

#import "UIColor+Variations.h"

@implementation UIColor (Variations)

- (void)extractRed:(CGFloat *)red
						 green:(CGFloat *)green
							blue:(CGFloat *)blue
						 alpha:(CGFloat *)alpha
{
  if ([self respondsToSelector:@selector(getRed:green:blue:alpha:)]) {
    BOOL worked = [self getRed:red green:green blue:blue alpha:alpha];
    
    if (NO == worked) {
      [self getWhite:red alpha:alpha];
      (*green) = (*red);
      (*blue)  = (*red);
    }
    
  } else {
    const CGFloat *components = CGColorGetComponents(self.CGColor);
    if (CGColorGetNumberOfComponents(self.CGColor) == 2) {
      (*red)   = (*green) = (*blue) = components[0];
      (*alpha) = components[1];
    } else {
      (*red)   = components[0];
      (*green) = components[1];
      (*blue)  = components[2];
      (*alpha) = components[3];
    }
  }
}

- (UIColor *)darkerColorByPercentage:(CGFloat)percentToBlack
{
  CGFloat red = 0.0, green = 0.0, blue = 0.0, alpha = 0.0;
	[self extractRed:&red green:&green blue:&blue alpha:&alpha];
	
	CGFloat multiplier = (1 - percentToBlack);
	return [UIColor colorWithRed:multiplier * red
												 green:multiplier * green
													blue:multiplier * blue
												 alpha:alpha];
}

- (UIColor *)invert
{
  CGFloat red = 0.0, green = 0.0, blue = 0.0, alpha = 0.0;
	[self extractRed:&red green:&green blue:&blue alpha:&alpha];
	
	return [UIColor colorWithRed:1 - red
												 green:1 - green
													blue:1 - blue
												 alpha:alpha];
}

- (BOOL)isDark
{
  CGFloat red = 0.0, green = 0.0, blue = 0.0, alpha = 0.0;
	[self extractRed:&red green:&green blue:&blue alpha:&alpha];
	
	return (red + green + blue) * alpha <= 1.75f; // mid-range colours are dark, too
}

@end

File: Sources/TripKitUI/helper/Categories/UIColor+Variations.swift
//
//  UIColor+Variations.swift
//  TripKitUI-iOS
//
//  Created by Adrian Schönig on 11/5/21.
//  Copyright © 2021 SkedGo Pty Ltd. All rights reserved.
//

import Foundation
import UIKit

extension UIColor {
  var rgba: (r: CGFloat, g: CGFloat, b: CGFloat, a: CGFloat) {
    var r: CGFloat = 0
    var g: CGFloat = 0
    var b: CGFloat = 0
    var a: CGFloat = 0
    if getRed(&r, green: &g, blue: &b, alpha: &a) {
      return (r, g, b, a)
    } else if let components = cgColor.components, components.count == 2 {
      return (components[0], components[0], components[0], components[1])
    } else {
      assertionFailure("Couldn't get colour")
      return (r, g, b, a)
    }
  }
  
  func darkerColor(percentage: CGFloat) -> UIColor {
    let multiplier = 1 - percentage
    let rgba = self.rgba
    return UIColor(red: multiplier * rgba.r,
                   green: multiplier * rgba.g,
                   blue: multiplier * rgba.b,
                   alpha: rgba.a)
  }
  
  public var isDark: Bool {
    let rgba = self.rgba
    return (rgba.r + rgba.g + rgba.b) * rgba.a <= 1.75 // mid-range colours are dark, too
  }
}
